Improve documentation about Kubernetes IngressRoute (#6058)
containous/traefik: docs/content/routing/providers/kubernetes-crd.md

|------------------------------------------|---------------------------------------------------------------|----------------------------------------------------------------|
| [IngressRoute](#kind-ingressroute)       | HTTP Routing                                                  | [HTTP router](../routers/index.md#configuring-http-routers)    |
| [Middleware](#kind-middleware)           | Tweaks the HTTP requests before they are sent to your service | [HTTP Middlewares](../../middlewares/overview.md)              |
| [TraefikService](#kind-traefikservice)   | Abstraction for HTTP loadbalancing/mirroring                  | [HTTP service](../services/index.md#configuring-http-services) |
| [IngressRouteTCP](#kind-ingressroutetcp) | TCP Routing                                                   | [TCP router](../routers/index.md#configuring-tcp-routers)      |
| [TLSOptions](#kind-tlsoption)            | Allows to configure some parameters of the TLS connection     | [TLSOptions](../../https/tls.md#tls-options)                   |

### Kind: `IngressRoute`

> `IngressRoute` is the CRD implementation of an HTTP [router](../routers/index.md#configuring-http-routers).

Reviewer: ```suggestion
`IngressRoute` is the CRD implementation of a [Traefik HTTP router](../routers/index.md#configuring-http-routers).
```

              httpOnly: true
              name: cookie
              secure: true
          strategy: RoundRobin
          weight: 10
      tls:
        certResolver: foo
        domains:
        - main: foo.com
          sans:
          - a.foo.com
          - b.foo.com
        options:
          name: opt
          namespace: default
        secretName: supersecret
    ```

    ```yaml tab="Middlewares"
    # All resources definition must be declared
    # Prefixing with /foo
    apiVersion: traefik.containo.us/v1alpha1
    kind: Middleware
    metadata:
      name: middleware1
      namespace: default
    spec:
      addPrefix:
        prefix: /foo
    ```

    ```yaml tab="TLSOption"
    apiVersion: traefik.containo.us/v1alpha1
    kind: TLSOption
    metadata:
      name: opt
      namespace: default
    
    spec:
      minVersion: VersionTLS12
    ```
    
    ```yaml tab="Secret"
    apiVersion: v1
    kind: Secret
    metadata:
      name: supersecret
    
    data:
      tls.crt: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCi0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0=
      tls.key: LS0tLS1CRUdJTiBQUklWQVRFIEtFWS0tLS0tCi0tLS0tRU5EIFBSSVZBVEUgS0VZLS0tLS0=
    ```

### Kind: `Middleware`

> To allow for the use of [middlewares](../../middlewares/overview.md) in an [IngressRoute](#kind-ingressroute), we defined a CRD for the `Middleware` kind.

Reviewer: ```suggestion
`Middleware` is the CRD implementation of a [Traefik "middleware"](../../middlewares/overview.md).
```

    kind: Middleware
    metadata:
      name: stripprefix
      namespace: foo
    
    spec:
      stripPrefix:
        prefixes:
          - /stripit
    ```
    
    ```yaml tab="IngressRoute"
    apiVersion: traefik.containo.us/v1alpha1
    kind: IngressRoute
    metadata:
      name: ingressroutebar
    
    spec:
      entryPoints:
        - web
      routes:
      - match: Host(`bar.com`) && PathPrefix(`/stripit`)
        kind: Rule
        services:
        - name: whoami
          port: 80
        middlewares:
        - name: stripprefix
          namespace: foo
    ```

!!! important "Cross-provider namespace"

	As Kubernetes also has its own notion of namespace, one should not confuse the kubernetes namespace of a resource
	(in the reference to the middleware) with the [provider namespace](../../middlewares/overview.md#provider-namespace),
	when the definition of the middleware comes from another provider.
	In this context, specifying a namespace when referring to the resource does not make any sense, and will be ignored.

More information about available middlewares in the dedicated [middlewares section](../../middlewares/overview.md).

### Kind: `TraefikService`

> To take advantage of all the capabilities of the Traefik [services](../services/index.md), we defined a CRD for the `TraefikService` kind.

Reviewer: ```suggestion
`TraefikService` is the CRD implementation of a ["Traefik Service"](../services/index.md).
```

> Once the `TraefikService` kind has been registered with the Kubernetes cluster,

Reviewer: ```suggestion
Register the `TraefikService` "kind" in the Kubernetes cluster before creating `TraefikService` objects, referencing services in the [`IngressRoute`](#kind-ingressroute)/[`IngressRouteTCP`](#kind-ingressroutetcp) objects or recursively in others `TraefikService` objects.
```

> it can then be used in `IngressRoute` definitions (as well as recursively in other Traefik Services).

Reviewer: ```suggestion
```

!!! info "Disambiguate Traefik and Kubernetes Services "

>     A `name` can refer either to a TraefikService or a Service, the `kind` field is used to break the ambiguity.

Reviewer: ```suggestion
    As the field `name` can reference different types of objects, use the field `kind` to avoid any ambiguity.
```

Reviewer: Avoid passive voice

>     The allowed values for this field are `TraefikService`, or `Service` (which is the default value and refers to a Kubernetes service).

Reviewer: ```suggestion

The field `kind` allows the following values:

* `Service` (default value): to reference a [Kubernetes Service](https://kubernetes.io/docs/concepts/services-networking/service/)
* `TraefikService`: to reference another [Traefik Service](../services/index.md)

```

             percent: 20
           - name: svc3
             kind: TraefikService
             percent: 20
    ```

    ```yaml tab="K8s Service"
    apiVersion: v1
    kind: Service
    metadata:
      name: svc1
      namespace: default
    
    spec:
      ports:
        - name: http
          port: 80
      selector:
        app: containous
        task: app1
    ---
    apiVersion: v1
    kind: Service
    metadata:
      name: svc2
      namespace: default
    
    spec:
      ports:
        - name: http
          port: 80
      selector:
        app: containous
        task: app2
    ```

!!! important "References and namespaces"

    If the optional `namespace` attribute is not set, the configuration will be applied with the namespace of the current resource.
    
    Additionally, when the definition of the `TraefikService` is from another provider,
    the cross-provider syntax (`service@provider`) should be used to refer to the `TraefikService`, just as in the middleware case.
    
    Specifying a namespace attribute in this case would not make any sense, and will be ignored (except if the provider is `kubernetescrd`).

### Kind `IngressRouteTCP`

> `IngressRouteTCP` is the CRD implementation of an TCP [router](../routers/index.md#configuring-tcp-routers).

Reviewer: ```suggestion
`IngressRouteTCP` is the CRD implementation of a [Traefik TCP router](../routers/index.md#configuring-tcp-routers).
```

> Once the `IngressRouteTCP` kind has been registered with the Kubernetes cluster, it can then be used as `IngressRouteTCP` object.

Reviewer: ```suggestion
Register the `IngressRouteTCP` "kind" in the Kubernetes cluster before creating `IngressRouteTCP` objects.
```

    
    spec:
      entryPoints:
        - footcp
      routes:
      # Match is the rule corresponding to an underlying router.
      - match: HostSNI(`*`)
        services:
        - name: foo
          port: 8080
          TerminationDelay: 400
          weight: 10
        - name: bar
          port: 8081
          TerminationDelay: 500
          weight: 10
      tls:
        certResolver: foo
        domains:
        - main: foo.com
          sans:
          - a.foo.com
          - b.foo.com
        options:
          name: opt
          namespace: default
        secretName: supersecret
        passthrough: false
    ```
    
    ```yaml tab="TLSOption"
    apiVersion: traefik.containo.us/v1alpha1
    kind: TLSOption
    metadata:
      name: opt
      namespace: default
    
    spec:
      minVersion: VersionTLS12
    ```
      
    ```yaml tab="Secret"
    apiVersion: v1
    kind: Secret
    metadata:
      name: supersecret
    
    data:
      tls.crt: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCi0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0=
      tls.key: LS0tLS1CRUdJTiBQUklWQVRFIEtFWS0tLS0tCi0tLS0tRU5EIFBSSVZBVEUgS0VZLS0tLS0=
    ```

### Kind: `TLSOption`

> `TLSOption` is the CRD implementation of [TLS Options](../../https/tls.md#tls-options).

Reviewer: ```suggestion
`TLSOption` is the CRD implementation of a [Traefik "TLS Option"](../../https/tls.md#tls-options).
```

> Once the TLSOption kind has been registered with the Kubernetes cluster or defined in the File Provider, it can then be referenced in `IngressRoute` or `IngressRouteTCP` definitions.

Reviewer: ```suggestion
Register the `TLSOption` "kind" in the Kubernetes cluster before creating `TLSOption` objects or referencing TLS options in the [`IngressRoute`](#kind-ingressroute) / [`IngressRouteTCP`](#kind-ingressroutetcp) objects.
```
